### DISCUSSION TITLE
how to build a model to predict node feature with only the same feature on neighbors?

### DISCUSSION THREAD

**Discussion by cappelchi** [danielegrattarola/spektral#326] [Q&A]
how to build a model to predict node feature with only the same feature on neighbors?

**Comment by danielegrattarola** [comment]
Hi, 

can you give more details as to what you would like to achieve?

**Comment by cappelchi** [comment]
for example, if I have the category of the nodes and adjacency matrix and I need to predict this category. I need to use category as a feature and as a label.

**Reply by cappelchi** [reply]
Synopsis: How to build a model, where the node features are both the input and the output for the model?

Details:
in my case, say, there is a node of interest and I'm going to predict its feature by analyzing the features of the neighborhood nodes. Therefore the neighborhood features are the 'input' at this moment, whereas the feature of the node of interest is the 'output' to be predicted. If we change the node of interest then the feature, which was the 'output' to be predicted is no longer the 'output', but the 'input' for the feature prediction in neighborhood nodes. 

Perhaps there are some examples or tutorials on this case?

**Reply by danielegrattarola** [reply]
It sound similar to label propagation from what I understand.
There are no tutorials for this specific use case, but you can have a look at this example

https://github.com/danielegrattarola/spektral/blob/master/examples/graph_prediction/custom_dataset.py

and create a graph dataset where the node features are the same as the labels. It's inefficient but it will allow you to use data loaders and the `model.fit()` approach. 

Alternatively, you can just manage the inputs/outputs of the model manually, if you write your own training and data management logic. 

Cheers

**Reply by cappelchi** [reply]
Salve Daniele,
I can't reproduce : https://github.com/danielegrattarola/spektral/blob/master/examples/graph_prediction/custom_dataset.py

ValueError: in user code:

    File "<ipython-input-43-c7cafe23fde0>", line 8, in train_step  *
        loss = loss_fn(target, predictions) + sum(model.losses)
    File "/usr/local/lib/python3.7/dist-packages/keras/losses.py", line 141, in __call__  **
        losses = call_fn(y_true, y_pred)
    File "/usr/local/lib/python3.7/dist-packages/keras/losses.py", line 245, in call  **
        return ag_fn(y_true, y_pred, **self._fn_kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/losses.py", line 1653, in categorical_crossentropy
        y_pred = tf.convert_to_tensor(y_pred)

    ValueError: None values not supported.

**Reply by cappelchi** [reply]
but works fine if I choose GeneralGNN.

**Reply by danielegrattarola** [reply]
What TF version are you on? 
I think this is an issue with topk pool, can you try removing it? I'll have a look into the problem asap

**Reply by cappelchi** [reply]
Python | 3.7.12 64bit [GCC 7.5.0]
IPython | 5.5.0
OS | Linux 5.4.144+ x86\_64 with Ubuntu 18.04 bionic
tensorflow | 2.7.0
keras | 2.7.0
Fri Feb 11 08:47:52 2022 UTC

**Reply by danielegrattarola** [reply]
Yeah just remove TopK pool until I have time to find a more decent fix, I have pushed a new version of the example that doesn't crash.

**Reply by cappelchi** [reply]
ok, grazie
